contracts/src/lib/LibGateway.sol

```diff
@@ -386,20 +400,149 @@ library LibGateway {
             supplySource = SupplySourceHelper.native();
         }
 
+        // If the crossnet destination is NOT the current network (network where the gateway is running),
+        // we add it to the postbox for further propagation.
+        // Even if we send for propagation, the execution of every message
+        // should increase the appliedNonce to allow the execution of the next message
+        // of the batch (this is way we have this after the nonce logic).
+        if (!crossMsg.to.subnetId.equals(s.networkName)) {
+            bytes32 cid = crossMsg.toHash();
+            s.postbox[cid] = crossMsg;
+            return;
+        }
+
+        // execute the message and get the receipt.
+        (bool success, bytes memory ret) = executeCrossMsg(crossMsg, supplySource);
+        if (success) {
+            sendReceipt(crossMsg, OutcomeType.Ok, ret);
+        } else {
+            sendReceipt(crossMsg, OutcomeType.ActorErr, ret);
+        }
+    }
+
+    /// @dev Execute the cross message using low level `call` method. This way ipc will
+    ///      catch contract revert messages as well. We need this because in `CrossMsgHelper.execute`
+    ///      there are `require` and `revert` calls, without reflexive call, the execution will
+    ///      revert and block the checkpoint submission process.
+    function executeCrossMsg(IpcEnvelope memory crossMsg, SupplySource memory supplySource) internal returns (bool success, bytes memory result) {
+        (success, result) = address(CrossMsgHelper).delegatecall(   // solhint-disable-line avoid-low-level-calls
+            abi.encodeWithSelector(CrossMsgHelper.execute.selector, crossMsg, supplySource)
+        );
+
+        if (success) {
+            return abi.decode(result, (bool, bytes));
+        }
+
+        return (success, result);
+    }
+
+    /// @notice Sends a receipt from the execution of a cross-message.
+    /// Only `Call` messages trigger a receipt. Transfer messages should be directly
+    /// handled by the peer client to return the funds to the from address in the
+    /// failing network.
+    /// (we could optionally trigger a receipt from `Transfer`s to, but without
+    /// multi-level execution it would be adding unnecessary overhead).
+    function sendReceipt(IpcEnvelope memory crossMsg, OutcomeType outcomeType, bytes memory ret) internal {
+        if (crossMsg.isEmpty()) {
+            revert CannotCreateIpcReceipt();
+        }
+
+        // if we get a `Receipt` do nothing, no need to send receipts.
+        // - And sending a `Receipt` to a `Receipt` could lead to amplification loops.
+        if (crossMsg.kind == IpcMsgKind.Result) {
+            return;
+        }
```

Comment:
I commented on the original PR, but recording here. How could we arrive here, and should this be an assertion instead?

Reply:
This would a sanity check. But a potential case would be:
- In the parent, a contract `Transfer` some value to another contract in the child.
- Child rejects the fund due to no `fallback` nor `receive`.
- Receipt send back to parent with refund.
- Original contract rejects fund transfer as no `fallback` nor `receive`. 

The above should reach here.